bokeh/model.py

```diff
@@ -739,9 +723,22 @@ def __exit__(self, type, value, traceback):
         for model in self._to_remove_after:
             model.document.remove_root(model)
 
-    def __enter__(self):
+    def __enter__(self, tried_already=False):
         for model in self._to_remove_after:
-            self._doc.add_root(model)
+            try:
+                self._doc.add_root(model)
+            except:
+                # see if some child of the model is in a doc, this is meant to
+                # handle a thing like:
+                #   p = figure()
+                #   box = HBox(children=[p])
+                #   show(box)
+                if tried_already is False:
+                    docs = [r.document for r in model.references()]
+                    if len(docs) == 1:
+                        self._doc = docs[0]
+                        self.__enter__(tried_already=True)
```

Comment:
Also if `len(docs) != 1` that's probably worth raising an explicit error stating the nature of the problem in terms of Bokeh concepts.